Add a Buffered Streamer around Espresso Streamer for batcher

Ayiga opened the pull request:
<!-- Closes #<ISSUE_NUMBER> -->
<!-- These comments should help create a useful PR message, please delete any remaining comments before opening the PR. -->
<!-- If there is no issue number make sure to describe clearly *why* this PR is necessary. -->
<!-- Mention open questions, remaining TODOs, if any -->

### This PR:
<!-- Describe what this PR adds to this repo and why -->
<!-- E.g. -->
<!-- * Implements feature 1 -->
<!-- * Fixes bug 3 -->
Adds a Buffer surrounding the `EspressoStreamer` for the Batcher to help lessen the cost / impact of the `Reset` calls that occur in the espressoBatchLoadingLoop.

<!-- ### This PR does not: -->
<!-- Describe what is out of scope for this PR, if applicable. Leave this section blank if it's not applicable -->
<!-- This section helps avoid the reviewer having to needlessly point out missing parts -->
<!-- * Implement feature 3 because that feature is blocked by Issue 4 -->
<!-- * Implement xyz because that is tracked in issue #123. -->
<!-- * Address xzy for which I opened issue #456 -->

<!-- ### Key places to review: -->
<!-- Describe key places for reviewers to pay close attention to -->
<!-- * file.rs, `add_integers` function -->
<!-- Or directly comment on those files/lines to make it easier for the reviewers -->

<!-- ### How to test this PR:  -->
<!-- Optional, uncomment the above line if this is relevant to your PR -->
<!-- If your PR is fully tested through CI there is no need to add this section -->
<!-- * E.g. `just test` -->

<!-- ### Things tested -->
<!-- Anything that was manually tested (that is not tested in CI). -->
<!-- E.g. building/running of docker containers. Changes to docker demo, ... -->
<!-- Especially mention anything untested, with reasoning and link an issue to resolve this. -->

<!-- Complete the following items before creating this PR -->
<!-- [ ] Issue linked or PR description mentions why this change is necessary. -->
<!-- [ ] PR description is clear enough for reviewers. -->
<!-- [ ] Documentation for changes (additions) has been updated (added).  -->
<!-- [ ] If this is a draft it is marked as "draft".  -->
<!-- [ ] For integration projects, make sure it won't break backward compatibility. -->

<!-- Details on maintaining backward compatibility for integration projects: -->
<!-- * Try to keep changes additive: add new, optional methods, flags, or parameters instead of modifying or removing existing functionality. -->
<!-- * If modification is necessary, it should be either a clear bug fix or guarded by a config/feature flag.  -->
<!-- * Follow Open Closed Principle and Interface Segregation Principle, clients should not be forced to depend on interfaces they do not use. -->

<!-- To make changes to this template edit https://github.com/EspressoSystems/.github/blob/main/PULL_REQUEST_TEMPLATE.md -->


---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1211175327473216

Ayiga committed:
Add a Buffered Streamer around Espresso Streamer for batcher

Because the `EspressoStreamer` is getting `Reset` during the Batcher process when
building a batch to submit to the `L1`, it hinders progress of the chain in a reasonable
amount of time, which ultimately causes it to stop creating non-empty blocks.

There are a number of factors that are contributing to this issue, but ultimately the
`Reset` is causing the `EspressoStreamer` to restart from `0` and it takes too long
before it catches back up to the next expected batch.

To remedy this, a Buffer can be used to mitigate this `Reset` and revert to a point
that is much closer to the desired next batch.

In testing it has been observed that the `SafeL2` can sometimes move backwards.  To
safe guard against this, it is better to `Reset` to the `FinalizedL2` position instead of
the `SafeL2` as this behavior has not been observed there.

QuentinI commented on espresso/interface.go:
```diff
@@ -0,0 +1,61 @@
+)
+
+// EspressoStreamerIFace defines the interface for the Espresso streamer.
+type EspressoStreamerIFace[B Batch] interface {
```
EspressoStreamerIFace is quite a mouthful...
Maybe we can name the interface EspressoStreamer and implementations Streamer and BufferedStreamer or something like that? That's what the SDK does with EspressoClient interface and Client and MultipleNodesClient implementors.

Ayiga replied:
This is fair.  I chose this route to minimize the Diff, otherwise it would result in a much larger diff.

Ayiga replied:
Renamed in 2927c80c68545daea02301774cc97f31564ae27d.

QuentinI commented on op-batcher/batcher/driver.go:
```diff
@@ -134,7 +134,7 @@ type BatchSubmitter struct {
 
 	espressoSubmitter *espressoTransactionSubmitter
-	espressoStreamer  espresso.EspressoStreamer[derive.EspressoBatch]
+	espressoStreamer  espresso.EspressoStreamerIFace[derive.EspressoBatch]
```
Is there a reason we don't want to just use a concrete impl here?

Ayiga replied:
The primary reason is to allow for the implementation to be swapped out (abstraction).  That way we can potentially make the `Buffered` version optional and `opt-in` in the future, or potentially make newer solutions that might be better\ without having to adjust this again.

QuentinI replied:
That's fair

QuentinI commented on espresso/interface.go:
```diff
@@ -0,0 +1,61 @@
+
+	// RemainingBatchesLen returns the number of pending undecided batches
+	// remaining in the streamer.
+	RemainingBatchesLen() int
```
As far as I can see, this is used only for a warning. I feel like we should just remove this method and only warn when we encounter an undecided batch inside the streamer code.

Ayiga replied:
This is fair.  I was just trying to keep the behavior of the code the same.

Ayiga replied:
Removed in 55dfadf4bfaab60c6aab4415ae5394b13fa0e7c3.

QuentinI commented on espresso/streamer.go:
```diff
@@ -433,16 +453,27 @@ func (s *EspressoStreamer[B]) HasNext(ctx context.Context) bool {
+		return false, err
 	}
 
+	shouldReset = hotshotState.BlockHeight < s.fallbackHotShotPos
```
Why don't we reset here directly?

Ayiga replied:
It seemed to be a better idea to me to keep the reset separate.  After all the original code doesn't reset here, but I wanted to surface more information that we had in this method to indicate whether or not we should Reset, that way the user could make that decision / adjustment themselves.

QuentinI replied:
Makes sense. In my personal opinion, it makes the code a bit harder to read, but it's not a hill I'd die on.

QuentinI commented on espresso/buffered_streamer.go:
```diff
@@ -0,0 +1,160 @@
+		if cast, castOk := b.streamer.(interface{ RefreshSafeL1Origin(eth.BlockID) error }); castOk {
+			// If the underlying streamer has a method to refresh the safe L1 origin,
+			// we call it to ensure it is aware of the new safe L1 origin.
+			return cast.RefreshSafeL1Origin(safeL1Origin)
```
Why do we not make this part of the interface?

Ayiga replied:
It was just in an effort to make the change minimal, and the interface easier to implement.  it can easily be added.

Ayiga replied:
Addressed in 0570d8230bf38087ff2c8cb11a8de86da06651e3.

QuentinI reviewed (commented):
Some nitpicks on code style, but implementation-wise LGTM

Ayiga committed:
Rename EspressoStreamer and EspressoStreamerIFace

Based on feedback provided by @QuentinI in PR review:
https://github.com/EspressoSystems/optimism-espresso-integration/pull/230#discussion_r2345028112

The name `EspressoStreamerIFace` is quite a long name, and the `IFace` suffix isn't
necessary since one could tell it's an `interface` by inspection, or using an `LSP`. The
feedback provided by @QuentinI suggested to rename `EspressoStreamerIFace` to
just `EspressoStreamer` so that it falls in line with our other code approaches.

This change renames `EspressoStreamer` to `BatchStreamer`

This change renames `EspressoStreamerIFace` to `EspressoStreamer`.

Ayiga committed:
Remove `RemainingBatchesLen` method

Based on feedback received from @QuentinI: https://github.com/EspressoSystems/optimism-espresso-integration/pull/230#discussion_r2345037174

The only reason `RemainingBatchesLen` exists is to serve as a check, and issue a
warning when things are running.  Even though this matches the existing behavior this
log seems to overlap with the logs corresponding to Undecided Batches which already
log warnings or errors.  As a result this method, and the log that utilizes it, seem to
be unnecessary and should be removed to eliminate noise.

This change removes `RemainingBatchesLen` and the uses of it

Ayiga committed:
Add `RefreshSafeL1Origin` to `EspressoStreamer` interface

Based on feedback provided by @QuentinI: https://github.com/EspressoSystems/optimism-espresso-integration/pull/230#discussion_r2345442127

Since the `RefreshSafeL1Origin` method can potentially be utilized in some places that
do not require a full `Refresh`, and for convenience, it makes sense to allow it to be a
separately exposed method distinct from `Refresh`.

This change adds `RefreshSafeL1Origin` as a required method in the
`EspressoStreamer` interface.

Ayiga committed:
Fix missed renamed references

Ayiga committed:
Rename enclave smoke test

Ayiga committed:
Fix refreshSafeL1Origin logic for Buffered Streamer

The buffered streamer is resetting its read position far more than it needs to, ultimately
reproducing the same issue that was already occurring with the unbuffered version.

In inspecting the behavior with a debugger, it seems we're resetting the reset position
unnecessarily when we receive the same safeL1Origin again.  Additionally, the logic
for determining the read position when the safeL1Origin advances also seems flawed,
in that it is very likely to reset too far in the past.  We really want to keep our relative
read position unless we're explicitly told to Reset.

This change addresses these issues in order to try and smooth out the batches being
returned, and avoid unnecessary reprocessing of previous batches.

QuentinI commented on espresso/buffered_streamer.go:
```diff
@@ -0,0 +1,153 @@
+// safe batch position.
+func (b *BufferedEspressoStreamer[B]) Reset() {
+	// Reset the buffered streamer state
+	b.readPos = 0
```
One thing I've noticed, shouldn't we be calling b.streamer.Reset() here as well?

Ayiga replied:
I don't think so.  That's kind of the point here, is that this wrapped Buffer is meant to mitigate the cost of the underlying Streamer being reset.  The way it achieves this is by allowing the Buffered Streamer to reset to a locally stored known safe L2 Block.

Without this mitigation here, the Buffer would be achieving nothing other than consuming memory.  The way it's currently implemented the underlying Streamer should only be `Reset` if the safe L2 reference ever moves backwards.

QuentinI replied:
Right, sorry, I kinda lost the context for a moment here 🤦‍♂️

Ayiga committed:
Add Unit Tests for BufferedEspressoStreamer

Fix BufferedEspressoStreamer behavior

While adding unit tests for the BufferedEspressoStreamer it was noticed that the
position of the L2 and the L1 for the Buffer were being mixed together at times.  This
would ultimately lead to very difficult to detect bugs based on observed behavior
alone.

With the addition of the unit tests identifying the issue, the buffer adjustment
behavior has been adjusted to apply to the L2 position in isolation away from the
L1 positions.  The L1 positions will cause a larger Reset in the underlying logic.
